pkg/distributor/distributor.go

```diff
@@ -350,10 +355,18 @@ func New(
 }
 
 func (d *Distributor) starting(ctx context.Context) error {
+	d.ingesterTaskWg.Add(d.cfg.PushWorkerCount)
+	for i := 0; i < d.cfg.PushWorkerCount; i++ {
+		go d.pushIngesterWorker()
+	}
 	return services.StartManagerAndAwaitHealthy(ctx, d.subservices)
 }
 
 func (d *Distributor) running(ctx context.Context) error {
+	defer func() {
+		close(d.ingesterTasks)
```

Comment:
I am not 100% sure, but I think it might be possible for this to cause a panic :thinking: If we are in the middle of a `Distributor.Push()` call when the context gets cancelled, we might close `d.ingesterTasks` here right before we try to add to it here: https://github.com/grafana/loki/blob/0738dd0dccd702df5bda992f6513eab830cf8a27/pkg/distributor/distributor.go#L653

Reply:
Change a bit the shutdown logic.